core/watcher/normalise.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package watcher
+
+import (
+	"github.com/juju/worker/v4"
+	"github.com/juju/worker/v4/catacomb"
+)
+
+// Normalise takes any watcher and normalises it down to a NotifyWatcher.
+// This is useful in legacy code that expects a NotifyWatcher.
+func Normalise[T any](source Watcher[T]) (NotifyWatcher, error) {
+	if w, ok := source.(NotifyWatcher); ok {
+		// If we are already a NotifyWatcher, we can return the source watcher.
+		return w, nil
+	}
+
+	ch := make(chan struct{})
+	w := &normaliseWatcher{
+		ch: ch,
+	}
+	loop := func() error {
+		defer close(ch)
+		for {
+			select {
+			case <-w.catacomb.Dying():
+				return w.catacomb.ErrDying()
+
+			case _, ok := <-source.Changes():
+				if !ok {
```

Comment:
We're not very consistent here - we either just return catacomb.ErrDying() unconditionally or some bespoke "watcher has been closed" error. I wonder if we should return any source watcher error if catacomb is not dying.

Reply:
If the source watcher exits with an error, then that error will be returned from the catacomb instead. Here this just deals with a race between the source watcher dying with success (and closing the channel).